### PR TITLE
[mixer] Introduce tracespan template into Mixer

### PR DESCRIPTION
**Pull request by douglas-reid**
**What this PR does / why we need it**:

This PR adds a template for trace spans to Mixer. A trace span template enables the integration of tracing adapters into Mixer. Early prototyping using this template demonstrated success in deriving tracing spans from proxy Report()s for services for jaeger, zipkin, and stackdriver backends. Operators can control the level of detail forwarded to tracing backends (as well as span names, for instance) using this template. The template also lays the groundwork for supporting Mixer ingestion of trace span data.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes https://github.com/istio/old_mixer_repo/issues/797, fixes: https://github.com/istio/old_mixer_repo/issues/1509

**Special notes for your reviewer**:

**Release note**:

```release-note
Tracing template added to Mixer
```

### CODE REVIEW

File: mixer/template/tracespan/tracespan.proto [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+// Example config:
+// ```
+// apiVersion: "config.istio.io/v1alpha2"
+// kind: tracespan
+// metadata:
+//   name: default
+//   namespace: istio-system
+// spec:
+//   traceId: request.headers["x-b3-traceid"]
+//   spanId: request.headers["x-b3-spanid"] | ""
```
Comment (guptasu) [inline mixer/template/tracespan/tracespan.proto]: are the x-b3-spanid and  x-b3-parentspanid well known header names. If we expect operators to use these for most case, consider explaining a bit about whey those tags mean.

File: mixer/template/tracespan/tracespan.proto [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+package traceSpan;
+
+import "google/protobuf/timestamp.proto";
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_REPORT;
+
+// TraceSpan represents an individual span within a distributed trace.
```
Comment (guptasu) [inline mixer/template/tracespan/tracespan.proto]: Should this be Trace span instead of TraceSpan?

Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: Are you suggesting renaming the package, or just wondering if I should put a space in between the two words?  This naming matches `LogEntry`.

File: mixer/template/tracespan/tracespan.proto [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+//
+// Example config:
+// ```
+// apiVersion: "config.istio.io/v1alpha2"
+// kind: tracespan
+// metadata:
+//   name: default
+//   namespace: istio-system
+// spec:
+//   traceId: request.headers["x-b3-traceid"]
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: I was under the impression we wanted "human case" config, where camelCase was preferred to snake_case.  But I actually am not certain.  @geeknoid ?

File: mixer/template/tracespan/tracespan.proto (hunk 1 above)
Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: these are well-known header names (from the zipkin set of headers).  they are described in full on our distributed tracing task. I've added a link to that Task here.

File: mixer/template/tracespan/tracespan.proto [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    // Required: This field MUST be configured in generated instances.
+    google.protobuf.Timestamp end_time = 6;
+
+    // Span tags are a set of <key, value> pairs that provide metadata for the
+    // entire span.
+    //
+    // Examples:
+    //   "http.method" : "POST"
+    //   "peer.service" : "foo.default.svc.cluster.local"
+    //   "request_size_bytes" : 234
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: I've removed the examples altogether, as they overlap with example attribute expressions.

File: mixer/template/tracespan/tracespan.proto [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    // and a line number where the operation is called. A best practice is to use
+    // the same display name within an application and at the same call point.
+    // This makes it easier to correlate spans in different traces.
+    //
+    // Required: This field MUST be configured in generated instances.
+    string span_name = 4;
+
+    // The start time of the span.
+    //
+    // Required: This field MUST be configured in generated instances.
```
[the remark this answers is not in the document]
Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: done.

File: mixer/template/tracespan/tracespan.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (geeknoid) [reply mixer/template/tracespan/tracespan.proto]: I think camelCase is what we're using.

File: mixer/template/tracespan/tracespan.proto (hunk 2 above)
Comment (guptasu) [reply mixer/template/tracespan/tracespan.proto]: I see, I was referring to the absence of space between the two words. Let's stick to LogEntry format then.

File: mixer/template/tracespan/tracespan.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (guptasu) [reply mixer/template/tracespan/tracespan.proto]: Thanks for clarifying.

File: mixer/template/tracespan/tracespan.proto (hunk 1 above)
Comment (guptasu) [reply mixer/template/tracespan/tracespan.proto]: Cool, that link will be helpful.

File: mixer/template/tracespan/tracespan.proto [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+// See also: [Distributed Tracing](https://istio.io/docs/tasks/telemetry/distributed-tracing.html)
+// for information on tracing within Istio.
+message Template {
+    // Trace ID is the unique identifier for a trace. All spans from the same
+    // trace share the same Trace ID.
+    //
+    // Required.
+    string trace_id = 1;
+
+    // Span ID is the unique identifier for a span within a trace. It is assigned
```
Comment (guptasu) [inline mixer/template/tracespan/tracespan.proto]: Is this auto assigned if not provided ?

Comment (douglas-reid) [reply mixer/template/tracespan/tracespan.proto]: Yes. Although, I think that is an adapter detail. Certainly, proper implementations will auto-assign.

**Pull request merged**